File: csharp/1163.last-substring-in-lexicographical-order.cs
/*
 * @lc app=leetcode id=1163 lang=csharp
 *
 * [1163] Last Substring in Lexicographical Order
 */

// @lc code=start
public class Solution {
    public string LastSubstring(string s) {
        int n = s.Length, i = 0, ni = 1, l = 0;
        while (ni + l < n) {
            char ci = s[i + l], cni = s[ni + l];
            if (ci == cni) {
                ++l;
                continue;
            }
            if (ci > cni) {
                ni = ni + l + 1;
                l = 0;
            } else {
                i = ni;
                ni++;
                l = 0;
            }
        }
        return s.Substring(i);
    }
}
// @lc code=end

File: java/1163.last-substring-in-lexicographical-order.java
/*
 * @lc app=leetcode id=1163 lang=java
 *
 * [1163] Last Substring in Lexicographical Order
 */

// @lc code=start
class Solution {
    public String lastSubstring(String s) {
        int n = s.length(), l = 0, ret = 0, cand = 1;
        while (cand + l < n) {
            int curr = s.charAt(ret + l), next = s.charAt(cand + l);
            if (curr == next) {
                ++l;
                continue;
            }
            if (curr > next) {
                cand = cand + l + 1;
                l = 0;
                continue;
            }
            ret = Math.max(cand, ret + l + 1);
            cand = ret + 1;
            l = 0;
        }
        return s.substring(ret);
    }
}
// @lc code=end

